central/metrics/aggregator/common/tracker_config.go

```diff
@@ -0,0 +1,165 @@
+	getters []LabelGetter[Finding], generator FindingGenerator[Finding],
+	gauge func(string, prometheus.Labels, int),
+) *TrackerConfig[Finding] {
+	return &TrackerConfig[Finding]{
+		category:    category,
+		description: description,
+		labelOrder:  makeLabelOrderMap(getters),
+		getters:     makeGettersMap(getters),
+		generator:   generator,
+		gauge:       gauge,
+	}
+}
+
+func (tc *TrackerConfig[Finding]) Reconfigure(ctx context.Context, registry *prometheus.Registry, filter string, cfg map[string]*storage.PrometheusMetricsConfig_Labels, period time.Duration) error {
+	mcfg, err := parseMetricLabels(cfg, tc.labelOrder)
+	if err != nil {
+		return err
+	}
+	q, err := search.ParseQuery(filter, search.MatchAllIfEmpty())
+	if err != nil {
+		return err
+	}
+	tc.SetMetricsConfiguration(q, mcfg)
+	// Force track only on reconfiguration, not on the initial tracker creation.
+	if tc.setPeriod(period) {
+		tc.track(ctx)
+	}
+	return tc.registerMetrics(registry, period)
+}
+
+// setPeriod returns true if tracker period has been reconfigured.
+func (tc *TrackerConfig[Finding]) setPeriod(period time.Duration) bool {
+	tc.metricsConfigMux.Lock()
+	defer tc.metricsConfigMux.Unlock()
+	if period > 0 {
+		if tc.ticker == nil {
+			tc.ticker = time.NewTicker(period)
+			return false
+		}
+		tc.ticker.Reset(period)
+		return true
+	}
+	if tc.ticker != nil {
+		tc.ticker.Stop()
+	}
+	return false
+}
+
+func (tc *TrackerConfig[Finding]) GetMetricsConfiguration() (*v1.Query, MetricsConfiguration) {
+	tc.metricsConfigMux.RLock()
+	defer tc.metricsConfigMux.RUnlock()
+	return tc.query, tc.metricsConfig
+}
+
+func (tc *TrackerConfig[Finding]) SetMetricsConfiguration(query *v1.Query, mcfg MetricsConfiguration) {
+	tc.metricsConfigMux.Lock()
+	defer tc.metricsConfigMux.Unlock()
+	tc.query = query
+	tc.metricsConfig = mcfg
+}
+
+func (tc *TrackerConfig[Finding]) registerMetrics(registry *prometheus.Registry, period time.Duration) error {
+	for metric, labelExpression := range tc.metricsConfig {
+		if err := metrics.RegisterCustomAggregatedMetric(string(metric), tc.description, period,
+			getMetricLabels(labelExpression, tc.labelOrder), registry); err != nil {
+			return fmt.Errorf("failed to register %s metric %q: %w", tc.category, metric, err)
+		}
+		if period > 0 {
+			log.Infof("Registered %s Prometheus metric %q", tc.category, metric)
+		}
+	}
+	if period == 0 {
+		log.Infof("Metrics collection is disabled for %s", tc.category)
+	}
+	return nil
+}
+
+func (tc *TrackerConfig[Finding]) track(ctx context.Context) {
+	query, mcfg := tc.GetMetricsConfiguration()
+	if len(mcfg) == 0 {
+		return
+	}
+	aggregator := makeAggregator(mcfg, tc.labelOrder)
+	for finding := range tc.generator(ctx, query, mcfg) {
+		aggregator.count(func(label Label) string {
+			return tc.getters[label](finding)
+		}, finding.Count())
+	}
+	for metric, records := range aggregator.result {
+		for _, rec := range records {
+			tc.gauge(string(metric), rec.labels, rec.total)
+		}
+	}
```

Comment:
I don't understand this concept. We make aggregators, we make them count things, and then later do nothing with them? Are they ephemeral only for those few lines?